### PR TITLE
[serve] Use Long Polling in Backend Worker

### PR DESCRIPTION
**Pull request by ijrsvt**
## Why are these changes needed?

Makes Serve Backend Workers Poll for Config Updates. This is part of the initiative started in #11905.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by simon-mo — APPROVED** [review]
Let's fix the comments and add a new tests as discussed offline

File: python/ray/serve/backend_worker.py [hunk 1]
```diff
@@ -369,6 +380,11 @@ def reconfigure(self, user_config) -> None:
...
             reconfigure_method(user_config)
 
+    async def _update_backend_configs(self, backend_configs):
+        for backend_tag, config in backend_configs.items():
+            if backend_tag == self.backend_tag:
+                self.update_config(config)
+
     def update_config(self, new_config: BackendConfig) -> None:
         self.config = new_config
         self.batch_queue.set_config(self.config.max_batch_size or 1,
```
[the remark this answers is not in the document]
Comment (ijrsvt) [reply python/ray/serve/backend_worker.py:390]: Ill also remove it from `RayServeWrappedReplica`

File: python/ray/serve/controller.py [hunk 2]
```diff
@@ -912,9 +911,10 @@ async def update_backend_config(
             self.notify_replica_handles_changed()
             self.notify_backend_configs_changed()
 
-            await self.broadcast_backend_config(backend_tag)
-
     async def broadcast_backend_config(self, backend_tag: BackendTag) -> None:
```
[the remark this answers is not in the document]
Comment (ijrsvt) [reply python/ray/serve/controller.py:914]: Done!

File: python/ray/serve/backend_worker.py [hunk 3]
```diff
@@ -141,11 +146,13 @@ def ensure_async(func: Callable) -> Callable:
     return sync_to_async(func)
 
 
+# Define Long-Poll Client
```
Comment (edoakes) [inline python/ray/serve/backend_worker.py:149]: accidental?

File: python/ray/serve/backend_worker.py [hunk 4]
```diff
@@ -369,6 +380,11 @@ def reconfigure(self, user_config) -> None:
                                          BACKEND_RECONFIGURE_METHOD)
             reconfigure_method(user_config)
 
+    async def _update_backend_configs(self, backend_configs):
+        for backend_tag, config in backend_configs.items():
+            if backend_tag == self.backend_tag:
```
Comment (edoakes) [inline python/ray/serve/backend_worker.py:385]: leave a note to remove this once we can poll for a specific key?

File: python/ray/serve/controller.py [hunk 5]
```diff
@@ -912,9 +911,10 @@ async def update_backend_config(
...
+
+        # Replace with notify_backend_configs_changed()
+        # After adding this to the Backend Replica
```
Comment (edoakes) [inline python/ray/serve/controller.py:917]: is this a comment or note to yourself?

File: python/ray/serve/tests/test_backend_worker.py [hunk 6]
```diff
@@ -111,15 +116,17 @@ def __call__(self, request):
...
+                                                mock_controller_with_name):
     class NonBatcher:
         def a(self, _):
             return "a"
 
         def b(self, _):
             return "b"
 
-    _ = await add_servable_to_router(NonBatcher, router)
+    _ = await add_servable_to_router(NonBatcher, router,
```
Comment (edoakes) [inline python/ray/serve/tests/test_backend_worker.py:128]: don't need the `_`

Comment (edoakes) [reply python/ray/serve/tests/test_backend_worker.py:128]: same for others

File: python/ray/serve/backend_worker.py (hunk 3 above)
Comment (ijrsvt) [reply python/ray/serve/backend_worker.py:149]: Note

File: python/ray/serve/controller.py (hunk 5 above)
Comment (ijrsvt) [reply python/ray/serve/controller.py:917]: The entire fn is deleted now :)

File: python/ray/serve/backend_worker.py (hunk 4 above)
Comment (ijrsvt) [reply python/ray/serve/backend_worker.py:385]: SGTM!

File: python/ray/serve/tests/test_backend_worker.py (hunk 6 above)
Comment (ijrsvt) [reply python/ray/serve/tests/test_backend_worker.py:128]: Done!

**Comment by ericl** [comment]
This is 60% flaky in master for serve:test_api: https://flakey-tests.ray.io

**Comment by edoakes** [comment]
Should be fixed here:
https://github.com/ray-project/ray/commit/7022278ce9b9830286f3747d3d070282bb27cea3

**Pull request merged**